statix.lang/trans/statix/lang/statics/core.stx

```diff
@@ -0,0 +1,567 @@
+        constraintOk(s, l),
+        constraintOk(s, r).
+    
+    constraintOk(s, CEqual(l, r, m)) :- {T1 T2}
+        T1 == termOk(s, l),
+        T2 == termOk(s, r), 
+        unifyTypes(T1, T2) | error $[Cannot unify types [T1] and [T2]].
+    
+    constraintOk(s, CArith(e1, _, e2, m)) :-
+        arithExpOk(s, e1),
+        arithExpOk(s, e2),
+        messageOk(s, m, RegularLocation()).
+        
+rules
+    termOk : scope * Term -> TERMTYPE
+    termsOk maps termOk(*, list(*)) = list(*)
+    
+    termOk(_, Int2Term(a@Int(_))) = INT() :-
+        @a.type := INT().
+    termOk(_, Str2Term(a@Str(_))) = STRING() :-
+        @a.type := STRING().
+    
+    termOk(_, Var2Term(Wld())) = UNKNOWN().
+    termOk(s, Var2Term(Var(var))) = typeOfVariable(s, var).
+    
+    termOk(_, List([])) = LIST(UNKNOWN()).
+    termOk(s, List(xs)) = LIST(T) :-
+        [T|_] == termsOk(s, xs).
+    
+    termOk(s, ListTail([], tail)) = LIST(T) :-
+        LIST(T) == termOk(s, tail).
+    termOk(s, ListTail(hs, tail)) = LIST(T) :-
+        [T|_] == termsOk(s, hs),
+        LIST(T) == termOk(s, tail).
+    
+    termOk(s, Tuple(t)) = TUPLE(TS) :-
+        TS == termsOk(s, t).
+    
+    termOk(s, Op(id, Terms2CommaTerms(ts))) = T :- {CTS sid len}
+        len == termLength(ts),
+        (CTS, sid) == typeOfConstructor(s, id, len),
+        T == typeOfSort(s, sid),
+        CTS == termsOk(s, ts).
+        
+    termOk(s, COp(cid, Terms2CommaTerms(ts))) = RT :- {ATS}
+        (ATS, RT) == typeOfConstraint(s, cid),
+        RT != UNIT() | error "Inline notation can only be used for functional constraints",
+        ATS == termsOk(s, ts).
+    
+    termOk(s, As(Wld(), term)) = termOk(s, term).
+    termOk(s, As(Var(var), term)) = T :-
+        T == typeOfVariable(s, var),
+        T == termOk(s, term).
+    
+    termOk(s, Ascribe(term, ty)) = T :-
+        T == termOk(s, term),
+        T == sortRefToType(s, ty).
+     
+    termOk(s, ArithOp(aExp)) = INT() :-
+        arithExpOk(s, aExp). 
+        
+rules
+  arithExpOk : scope * ArithExp
+  
+  arithExpOk(_, Int2ArithExp(_)).
+  
+  arithExpOk(_, Var2ArithExp(Wld())).
+  arithExpOk(s, Var2ArithExp(Var(var))) :- 
+    typeOfVariable(s, var) == INT().
+    
+  arithExpOk(s, AAdd(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).
+    
+  arithExpOk(s, AMul(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).
+    
+  arithExpOk(s, ASub(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).
+    
+  arithExpOk(s, AMin(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).
+    
+  arithExpOk(s, AMax(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).    
+    
+  arithExpOk(s, AMod(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).
+    
+  arithExpOk(s, ADiv(e1, e2)) :-
+    arithExpOk(s, e1),
+    arithExpOk(s, e2).
+
+signature
+  sorts MessageLocation constructors
+    RegularLocation : MessageLocation
+    TryLocation : MessageLocation
```

Comment:
I think that, instead of this `MessageLocation`, there should be two constraints:
```
messageOk(s, Message(kind, content, origin), loc) :-
    kind == ERROR() | error $[/* use 'try' */],
    messageContentOk(s, content).
    
tryMessageOk(s, Message(kind, content, origin), loc) :-
    messageContentOk(s, content).
```